core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -432,6 +435,57 @@ private[spark] class Executor(
   }
 
   /**
+   * Supervises the killing / cancellation of a task by sending the interrupted flag, optionally
+   * sending a Thread.interrupt(), and monitoring the task until it finishes.
+   */
+  private class TaskReaper(taskRunner: TaskRunner, interruptThread: Boolean) extends Runnable {
+
+    private[this] val killPollingFrequencyMs: Long =
+      conf.getTimeAsMs("spark.task.killPollingFrequency", "10s")
+
+    private[this] val killTimeoutMs: Long = conf.getTimeAsMs("spark.task.killTimeout", "2m")
+
+    private[this] val takeThreadDump: Boolean =
+      conf.getBoolean("spark.task.threadDumpKilledTasks", true)
+
+    override def run(): Unit = {
+      val startTimeMs = System.currentTimeMillis()
+      def elapsedTimeMs = System.currentTimeMillis() - startTimeMs
+
+      while (!taskRunner.isFinished && elapsedTimeMs < killTimeoutMs) {
+        taskRunner.kill(interruptThread = interruptThread)
+        taskRunner.synchronized {
+          Thread.sleep(killPollingFrequencyMs)
+        }
+        if (!taskRunner.isFinished) {
+          logWarning(s"Killed task ${taskRunner.taskId} is still running after $elapsedTimeMs ms")
+          if (takeThreadDump) {
+            try {
+              val threads = Utils.getThreadDump()
+              threads.find(_.threadName == taskRunner.threadName).foreach { thread =>
+                logWarning(s"Thread dump from task ${taskRunner.taskId}:\n${thread.stackTrace}")
+              }
+            } catch {
+              case NonFatal(e) =>
+                logWarning("Exception thrown while obtaining thread dump: ", e)
+            }
+          }
+        }
+      }
+      if (!taskRunner.isFinished && killTimeoutMs > 0 && elapsedTimeMs > killTimeoutMs) {
+        if (isLocal) {
+          logError(s"Killed task ${taskRunner.taskId} could not be stopped within " +
```

Comment:
Even if we did throw an exception here, it wouldn't exit the JVM in local mode because we don't set an uncaught exception handler in local mode (see code further up in this file).